File: problems/2022_09_24/0078/Solution.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        self.res = []
        self.dfs(nums, [], 0)
        return self.res
    
    def dfs(self, nums, curr, index):
        if index >= len(nums):
            self.res.append(curr.copy())
            return
        
        # decision to include current index
        curr.append(nums[index])
        self.dfs(nums, curr, index + 1)
        
        # decision not to include current index
        curr.pop()
        self.dfs(nums, curr, index + 1)

File: problems/2022_09_24/0078/Solution.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> list = new ArrayList<>();
        backtrack(0, res, list, nums);
        return res;
    }

    private void backtrack(int i, List<List<Integer>> res, List<Integer> list, int[] nums) {
        if (i >= nums.length) {
            res.add(new ArrayList<>(list));
            return;
        }
        // add the element to the list
        list.add(nums[i]);
        backtrack(i + 1, res, list, nums);

        // do not add the element to the list
        list.remove(list.size() - 1);
        backtrack(i + 1, res, list, nums);
    }
}